flink-runtime/src/main/java/org/apache/flink/runtime/state/DuplicatingCheckpointOutputStream.java

```diff
@@ -0,0 +1,287 @@
+		CheckpointStreamFactory.CheckpointStateOutputStream secondaryOutputStream) throws IOException {
+		this(primaryOutputStream, secondaryOutputStream, DEFAULT_BUFFER_SIZER);
+	}
+
+	public DuplicatingCheckpointOutputStream(
+		CheckpointStreamFactory.CheckpointStateOutputStream primaryOutputStream,
+		CheckpointStreamFactory.CheckpointStateOutputStream secondaryOutputStream,
+		int bufferSize) throws IOException {
+
+		this.primaryOutputStream = Preconditions.checkNotNull(primaryOutputStream);
+		this.secondaryOutputStream = Preconditions.checkNotNull(secondaryOutputStream);
+
+		this.buffer = new byte[bufferSize];
+		this.bufferIdx = 0;
+
+		this.secondaryStreamException = null;
+
+		checkForAlignedStreamPositions();
+	}
+
+	@Override
+	public void write(int b) throws IOException {
+
+		if (buffer.length <= bufferIdx) {
+			flushInternalBuffer();
+		}
+
+		buffer[bufferIdx] = (byte) b;
+		++bufferIdx;
+	}
+
+	@Override
+	public void write(byte[] b) throws IOException {
+
+		write(b, 0, b.length);
+	}
+
+	@Override
+	public void write(byte[] b, int off, int len) throws IOException {
+
+		if (buffer.length <= len) {
+
+			flushInternalBuffer();
+			writeThroughInternal(b, off, len);
+		} else {
+
+			if (buffer.length < len + bufferIdx) {
+				flushInternalBuffer();
+			}
+
+			System.arraycopy(b, off, buffer, bufferIdx, len);
+			bufferIdx += len;
+		}
+	}
+
+	@Override
+	public long getPos() throws IOException {
+		final long referencePos = primaryOutputStream.getPos();
+		return referencePos + bufferIdx;
+	}
+
+	@Override
+	public void flush() throws IOException {
+
+		flushInternalBuffer();
+		primaryOutputStream.flush();
+
+		if (secondaryStreamException == null) {
+			try {
+				secondaryOutputStream.flush();
+			} catch (IOException flushEx) {
+				handleSecondaryStreamOnException(flushEx);
+			}
+		}
+	}
+
+	@Override
+	public void sync() throws IOException {
+
+		flushInternalBuffer();
+		primaryOutputStream.sync();
+
+		if (secondaryStreamException == null) {
+			try {
+				secondaryOutputStream.sync();
+			} catch (IOException syncEx) {
+				handleSecondaryStreamOnException(syncEx);
+			}
+		}
+	}
+
+	@Override
+	public void close() throws IOException {
+
+		IOException exCollector = null;
+
+		try {
+			flushInternalBuffer();
+		} catch (IOException flushEx) {
+			exCollector = flushEx;
+		}
+
+		try {
+			primaryOutputStream.close();
+		} catch (IOException closeEx) {
+			exCollector = ExceptionUtils.firstOrSuppressed(closeEx, exCollector);
+		}
+
+		if (secondaryStreamException == null) {
+			try {
+				secondaryOutputStream.close();
+			} catch (IOException closeEx) {
+				handleSecondaryStreamOnException(closeEx);
+			}
+		}
+
+		if (exCollector != null) {
+			throw exCollector;
+		}
+	}
+
+	private void checkForAlignedStreamPositions() throws IOException {
+
+		final long primaryPos = primaryOutputStream.getPos();
+
+		if (secondaryStreamException != null) {
+			return;
```

Comment:
This `if` could place before `final long primaryPos = primaryOutputStream.getPos();`

Reply:
👍